core/controllers/classifier.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controllers for communicating with the VM for training classifiers."""
+
+import hashlib
+import hmac
+import json
+
+from core.controllers import base
+from core.domain import classifier_domain
+from core.domain import classifier_services
+from core.domain import config_domain
+from core.domain import exp_services
+
+import feconf
+
+
+def validate_request(handler):
+    """Decorator that checks if the incoming request for storing trained
+    classifier is valid."""
+    def test_is_valid(self, **kwargs):
+        message = self.request.get('payload')
+        message = json.loads(message)
+        signature = message.get('signature')
+        vm_id = message.get('vm_id')
+        message.pop('signature')
+        message = json.dumps(message, sort_keys=True)
+        secret = str([val['shared_secret_key'] for val in (
+            config_domain.VMID_SHARED_SECRET_KEY_MAPPING.value) if val[
+                'vm_id'] == vm_id][0])
+        generated_signature = hmac.new(
+            secret, message, digestmod=hashlib.sha256).hexdigest()
+        if generated_signature != signature:
+            raise self.UnauthorizedUserException
```

Comment:
401 is an authorization error, shown when a user doesn't have permission to visit the page.

In this case it's a bit unclear, since the user is a machine. I think either 400 or 401 is fine. I don't have a strong opinion.